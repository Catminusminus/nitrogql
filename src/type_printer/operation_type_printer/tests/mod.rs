use insta::assert_snapshot;

use crate::{
    ast::{OperationDocument, TypeSystemDocument},
    extension_resolver::resolve_extensions,
    graphql_builtins::generate_builtins,
    graphql_parser::parser::{parse_operation_document, parse_type_system_document},
    source_map_writer::just_writer::JustWriter,
};

use super::{QueryTypePrinter, QueryTypePrinterOptions};

fn type_system() -> TypeSystemDocument<'static> {
    let mut doc = parse_type_system_document(
        "
            type User implements HasID {
                id: ID!
                name: String!
                type: UserType!
                age: Int
                posts: [HasID!]!
            }
            interface HasID {
                id: ID!
            }
            enum UserType {
                NormalUser
                PremiumUser
            }
            type Bot implements HasID {
                id: ID!
            }

            type Post implements HasID {
                id: ID!
                title: String!
                tags: [String!]
                body: String!
            }

            type Tweet implements HasId {
                id: ID!
                body: String!
            }

            input UserSearchQuery {
                age: Int
                name: String
                keywords: [String!]
            }

            type Query {
                me: User!
            }
            ",
    )
    .unwrap();
    doc.extend(generate_builtins());
    let doc = resolve_extensions(doc).unwrap();
    doc
}

#[test]
fn basic_type_printing() {
    let doc = parse_operation_document(
        "
        query {
            me {
                id name type age
            }
        }
        ",
    )
    .unwrap();
    let printed = print_document(&doc);
    assert_snapshot!(printed);
}

#[test]
fn fragment_spread() {
    let doc = parse_operation_document(
        "
        query {
            me {
                id name type age
                posts {
                    id
                    ...F
                }
            }
        }
        fragment F on HasId {
            id
        }
        ",
    )
    .unwrap();
    let printed = print_document(&doc);
    assert_snapshot!(printed);
}

#[test]
fn fragment_inline_spread() {
    let doc = parse_operation_document(
        "
        query {
            me {
                id name type age
                posts {
                    ... {
                        id
                    }
                }
            }
        }
        ",
    )
    .unwrap();
    let printed = print_document(&doc);
    assert_snapshot!(printed);
}

fn print_document(document: &OperationDocument) -> String {
    let mut result = String::new();
    let mut writer = JustWriter::new(&mut result);
    let mut printer = QueryTypePrinter::new(QueryTypePrinterOptions::default(), &mut writer);
    printer.print_document(document);
    result
}
