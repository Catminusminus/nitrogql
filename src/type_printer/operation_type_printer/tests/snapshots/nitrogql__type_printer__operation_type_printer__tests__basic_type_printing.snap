---
source: src/type_printer/operation_type_printer/tests/mod.rs
expression: printed
---
import type { TypedDocumentNode } from "@graphql-typed-document-node/core";
import type * as Schema from "";

type Query = Schema.__SelectionSet<Schema.Query, {
  me: Schema.__SelectionSet<Schema.User, {
    id: Schema.ID;
    name: Schema.String;
    type: Schema.UserType;
    age: Schema.Int | null;
  }>;
}>;

type Variables = {};

export const Query: TypedDocumentNode<Query, Variables>;


